### PR TITLE
Use coreclr msbuild

### PR DESCRIPTION
**Pull request by agocke**
This change ports us to use the netcore version of MSBuild running on CoreCLR instead of running MSBuild on Mono.

The primary change is the addition of the MSbuildToolset xproj, which is a netcore-targeted project that assembles the pieces for MSBuild running on CoreCLR. The rest of the pieces, like reference assemblies and assorted third-party tasks+targets, are assembled by the `build_toolset.ps1` script.

Running the `build_toolset.ps1` script builds an OS X and Linux version of the toolset directories, which are each tar-bzipped and uploaded to Azure for use on their respective platforms.

### CODE REVIEW

**Comment by agocke** [comment]
@dotnet-bot retest prtest/lin/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_lin_dbg_unit32/4986/
// Retest reason:

**Comment by agocke** [comment]
@dotnet-bot retest prtest/lin/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_lin_dbg_unit32/4986/
// Retest reason:

**Comment by agocke** [comment]
@jaredpar @TyOverby @dotnet/roslyn-compiler @dotnet/roslyn-infrastructure I think this is ready for review now.

File: .gitattributes [hunk 1]
```diff
@@ -3,6 +3,10 @@
 ###############################################################################
 * text=auto encoding=UTF-8
 
+# csc/vbc are shell scripts and should always have unix line endings
+src/Compilers/CSharp/CscCore/csc text eol=lf
+src/Compilers/VisualBasic/VbcCore/vbc text eol=lf
```
Comment (jaredpar) [inline .gitattributes]: Can you elaborate why this was necessary here? There are several other shell scripts in the repo that aren't listed here.  Do they suffer from whatever issue prompted this change?

File: build/MSBuildToolset/Program.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+﻿
```
Comment (jaredpar) [inline build/MSBuildToolset/Program.cs]: Copyright

File: Makefile [hunk 3]
```diff
@@ -1,47 +1,52 @@
 SHELL = /usr/bin/env bash
 OS_NAME = $(shell uname -s)
-NUGET_PACKAGE_NAME = nuget.67
+NUGET_PACKAGE_NAME = nuget.68
 BUILD_CONFIGURATION = Debug
-BOOTSTRAP_PATH = $(shell pwd)/Binaries/Bootstrap
+TOOLSET_TMP_PATH = $(shell pwd)/toolset
```
Comment (jaredpar) [inline Makefile]: Can we put this under Binaries as well?  That way all of our build artifacts remain under a single root directory: Binaries.

File: Makefile [hunk 4]
```diff
@@ -1,47 +1,52 @@
...
+BINARIES_PATH = $(shell pwd)/Binaries
+BOOTSTRAP_PATH = $(BINARIES_PATH)/Bootstrap
 BUILD_LOG_PATH =
 XUNIT_VERSION = 2.1.0
+HOME_DIR = $(shell cd ~ && pwd)
```
Comment (jaredpar) [inline Makefile]: Man I missed good old shell scripting tricks ...

File: .gitattributes (hunk 1 above)
Comment (agocke) [inline .gitattributes]: Right. The issue is that these shell scripts are included in the toolset packages. Normally, the shell scripts in our repo are only run by cloning onto a Linux/Mac machine, and git automatically chooses LF as the line ending.

However, right now the toolset packages must be built on Windows, and so the files are cloned with CRLF and then copied into the unix packages with those line endings.

When we run them on a target unix OS, the files then blow up with bad line endings.

File: Makefile [hunk 5]
```diff
@@ -1,47 +1,52 @@
...
 endif
 
-MONO_PATH = /tmp/$(MONO_TOOLSET_NAME)/bin/mono
-MSBUILD_ADDITIONALARGS := $(MSBUILD_ADDITIONALARGS) /p:MonoToolsetPath=$(MONO_PATH)
-TOOLSET_ARGS = $(MSBUILD_ADDITIONALARGS) /p:CscToolPath=$(ROSLYN_TOOLSET_PATH) /p:CscToolExe=csc /p:VbcToolPath=$(ROSLYN_TOOLSET_PATH) /p:VbcToolExe=vbc
+MSBUILD_CMD = $(ROSLYN_TOOLSET_PATH)/corerun $(ROSLYN_TOOLSET_PATH)/MSBuild.exe $(MSBUILD_ARGS)
 
 all: toolset
-	$(MONO_PATH) ~/.nuget/packages/Microsoft.Build.Mono.Debug/14.1.0/lib/MSBuild.exe $(TOOLSET_ARGS) CrossPlatform.sln
+	export ReferenceAssemblyRoot=$(ROSLYN_TOOLSET_PATH)/reference-assemblies/Framework ; \
```
Comment (jaredpar) [inline Makefile]: In general environment variables have all upper case names on Windows.  Is that a thing for Unix as well?  Or is normal cased names fine here?

File: Makefile [hunk 6]
```diff
@@ -1,47 +1,52 @@
...
+	$(MSBUILD_CMD) CrossPlatform.sln
 
 bootstrap: toolset
-	$(MONO_PATH) ~/.nuget/packages/Microsoft.Build.Mono.Debug/14.1.0/lib/MSBuild.exe $(TOOLSET_ARGS) src/Compilers/CSharp/CscCore/CscCore.csproj ; \
-	$(MONO_PATH) ~/.nuget/packages/Microsoft.Build.Mono.Debug/14.1.0/lib/MSBuild.exe $(TOOLSET_ARGS) src/Compilers/VisualBasic/VbcCore/VbcCore.csproj ; \
-	mkdir -p $(BOOTSTRAP_PATH) ; \
-	cp Binaries/$(BUILD_CONFIGURATION)/csccore/* $(BOOTSTRAP_PATH) ; \
-	cp Binaries/$(BUILD_CONFIGURATION)/vbccore/* $(BOOTSTRAP_PATH) ; \
-	build/scripts/crossgen.sh $(BOOTSTRAP_PATH) ;
+	export ReferenceAssemblyRoot=$(ROSLYN_TOOLSET_PATH)/reference-assemblies/Framework ; \
```
Comment (jaredpar) [inline Makefile]: Here be dragons ... :dragon:

File: Makefile (hunk 5 above)
Comment (agocke) [inline Makefile]: We don't get a choice here -- this exact environment variable is the one MSBuild is looking for in its utilities task.

File: Makefile [hunk 7]
```diff
@@ -50,25 +55,22 @@ test:
...
-/tmp/$(NUGET_PACKAGE_NAME).zip:
-	@pushd /tmp/ ; \
-	curl -O https://dotnetci.blob.core.windows.net/roslyn/$(NUGET_PACKAGE_NAME).zip ; \
-	unzip -uoq $(NUGET_PACKAGE_NAME).zip -d ~/ ; \
-	chmod +x ~/.nuget/packages/Microsoft.Build.Mono.Debug/14.1.0/lib/MSBuild.exe
+	@rm -rf $(TOOLSET_TMP_PATH)
+
+toolset: $(TOOLSET_TMP_PATH)/$(ROSLYN_TOOLSET_NAME) $(TOOLSET_TMP_PATH)/$(NUGET_PACKAGE_NAME).zip
+
+$(TOOLSET_TMP_PATH)/$(ROSLYN_TOOLSET_NAME): | $(TOOLSET_TMP_PATH)
```
Comment (jaredpar) [inline Makefile]: What does the `|` represent here?

Comment (agocke) [inline Makefile]: This is a `make` order-only dependency: https://www.gnu.org/software/make/manual/html_node/Prerequisite-Types.html

File: build/MSBuildToolset/Program.cs [hunk 8]
```diff
@@ -0,0 +1,10 @@
...
+namespace DoNotUse
+{
+    public class Program
+    {
+        public static void Main(string[] args)
+        {
```
Comment (jaredpar) [inline build/MSBuildToolset/Program.cs]: Pedantic nit: if you don't want anyone to use it ...

``` csharp
public static int Main() => 1;
```

File: build/MSBuildToolset/build_toolset.ps1 [hunk 9]
```diff
@@ -0,0 +1,71 @@
+﻿# Builds the toolset for use in unix msbuild-coreclr
```
Comment (jaredpar) [inline build/MSBuildToolset/build_toolset.ps1]: all of our powershell scripts use `-` instead of `_`.

File: build/Targets/GenerateAssemblyInfo.targets [hunk 10]
```diff
@@ -68,21 +68,11 @@
 
     <!-- The necessity of this Task is tracked by https://github.com/dotnet/roslyn/issues/8421 -->
     <WriteCodeFragmentEx 
-        Condition="'$(OS)' == 'Windows_NT'"
```
Comment (jaredpar) [inline build/Targets/GenerateAssemblyInfo.targets]: Nice ... this means we can get deterministic builds on Linux / Mac now too!

File: build/Targets/Settings.targets [hunk 11]
```diff
@@ -50,6 +49,12 @@
     <CopyVsixManifestToOutput>false</CopyVsixManifestToOutput>
   </PropertyGroup>
 
+  <!-- Point to the reference assemblies on unix -->
+  <PropertyGroup Condition="'$(OS)' != 'Windows_NT'">
+      <TargetFrameworkRootPath>$(MSBuildBinPath)\reference-assemblies\Framework</TargetFrameworkRootPath>
+  </PropertyGroup>
+
+
```
Comment (jaredpar) [inline build/Targets/Settings.targets]: Double blank line.

File: build/Targets/Settings.targets [hunk 12]
```diff
@@ -97,6 +102,10 @@
     <UseRoslynAnalyzers Condition="'$(UseRoslynAnalyzers)' == ''">true</UseRoslynAnalyzers>
 
     <Deterministic>True</Deterministic>
+
+    <!-- Only portable PDBs are supported xplat -->
+    <DebugType Condition="'$(OS)' != 'Windows_NT'">portable</DebugType>
+    <DebugType Condition="'$(OS)' == 'Windows_NT'">pdbonly</DebugType>
```
Comment (jaredpar) [inline build/Targets/Settings.targets]: It's all happening!!!  

CC @tmat our Linux / Mac jobs now going to produce portable PDB on every run.

File: src/Compilers/Core/MSBuildTask/Portable/BuildClientShim.cs [hunk 13]
```diff
@@ -16,9 +16,6 @@ internal static class BuildClientShim
             BuildPaths buildPaths,
             string keepAlive,
             string libEnvVariable,
-            CancellationToken cancellationToken)
-        {
-            throw new NotSupportedException();
-        }
+            CancellationToken cancellationToken) => Task.FromResult<BuildResponse>(null);
```
Comment (jaredpar) [inline src/Compilers/Core/MSBuildTask/Portable/BuildClientShim.cs]: Why this change?  I updated this code path to expliictly not use `null` anymore.

Comment (agocke) [inline src/Compilers/Core/MSBuildTask/Portable/BuildClientShim.cs]: There's no catch around the calling code though -- this currently crashes the build.

Comment (jaredpar) [inline src/Compilers/Core/MSBuildTask/Portable/BuildClientShim.cs]: Oh ... that change is only in my private branch.  I've never pushed that.

File: .gitattributes (hunk 1 above)
Comment (jaredpar) [inline .gitattributes]: Explanation sounds good, please add it as a comment :smile:

**Comment by jaredpar** [comment]
:+1:  modulo a few outstanding comments.

File: build/Targets/Settings.targets (hunk 12 above)
Comment (tmat) [inline build/Targets/Settings.targets]: Yay! :)

File: build/Targets/GenerateCompilerInternals.targets [hunk 14]
```diff
@@ -35,7 +35,7 @@
   </PropertyGroup>
 
   <PropertyGroup>
-	<MonoPrefix Condition="'$(OS)' != 'Windows_NT'">$(MonoToolsetPath) </MonoPrefix>
+      <MonoPrefix Condition="'$(OS)' != 'Windows_NT'">mono</MonoPrefix>
```
Comment (TyOverby) [inline build/Targets/GenerateCompilerInternals.targets]: Why do we still have this line? Is MonoPrefix still required?

Comment (agocke) [inline build/Targets/GenerateCompilerInternals.targets]: Yes, just to run the generators. Converting the generators over to netcore is my next task.

File: build/MSBuildToolset/build-toolset.ps1 [hunk 15]
```diff
@@ -0,0 +1,71 @@
+﻿# Builds the toolset for use in unix msbuild-coreclr
```
Comment (jaredpar) [inline build/MSBuildToolset/build-toolset.ps1]: Can we change the powershell script to have the following pattern?

``` ps1
$ErrorActionPreference="Stop"

try
{
   # Code that you have today 
catch
{
    write-host "Error: $($_.Exception.Message)"
    exit 1
}
```

The issue with powershell is that it essentially does no error checking.  If `dotnet` failed to exist or restore simply failed the script would print an error message and continue on.  Eventually exiting with almost certainly a 0 exit code indicating success.  The above pattern helps lock down powershell scripts and prevent this from happening.

Comment (agocke) [inline build/MSBuildToolset/build-toolset.ps1]: Sure, will fix.

**Comment by agocke** [comment]
@TyOverby Was that a :+1:? :smile:

File: build/Targets/GenerateCompilerInternals.targets (hunk 14 above)
Comment (TyOverby) [inline build/Targets/GenerateCompilerInternals.targets]: gotcha

**Comment by TyOverby** [comment]
LGTM
